### PR TITLE
[TorchComms] Support training with EP

### PR DESCRIPTION
**Pull request by mori360**
support _build_mesh_with_ep

`TEST_BACKEND=nccl TRAIN_FILE=torchtitan.experiments.torchcomms.train CONFIG_FILE="./torchtitan/models/qwen3/train_configs/qwen3_moe_debug.toml" ./run_train.sh  --parallelism.expert_parallel_degree 2`

[rank0]:[titan] 2025-10-16 17:32:31,142 - root - INFO - Building qwen3 debugmodel_moe with Qwen3ModelArgs(_enforced='This field is used to enforce all fields have defaults.', dim=256, n_layers=8, n_heads=16, n_kv_heads=8, vocab_size=2048, head_dim=128, hidden_dim=3072, norm_eps=1e-06, rope_theta=1000000, qk_norm=True, max_seq_len=4096, depth_init=True, use_flex_attn=False, attn_mask_type='causal', eos_id=151645, enable_weight_tying=False, moe_enabled=True, moe_inter_dim=768, moe_args=MoEArgs(num_experts=64, num_shared_experts=0, score_func='softmax', route_norm=True, route_scale=1.0, score_before_experts=False, top_k=8, use_grouped_mm=True, load_balance_coeff=0.001, _debug_force_load_balance=False))
...
[rank0]:[titan] 2025-10-16 17:32:40,167 - root - INFO - step:  1  loss:  8.1372  grad_norm:  2.8767  memory:  4.90GiB(5.16%)  tps: 1,821  tflops: 0.74  mfu: 0.07%
[rank0]:[titan] 2025-10-16 17:32:40,167 - root - INFO - Synchronizing and adjusting timeout for all ProcessGroups to 0:01:40
[rank0]:/data/users/yifanmao/pytorch/torch/distributed/distributed_c10d.py:1543: UserWarning: Set timeout is now only supported for either nccl or gloo.
[rank0]:  warnings.warn("Set timeout is now only supported for either nccl or gloo.")
[rank0]:[titan] 2025-10-16 17:32:40,371 - root - INFO - step:  2  loss:  7.3916  grad_norm:  3.0698  memory:  4.91GiB(5.17%)  tps: 80,530  tflops: 32.75  mfu: 3.31%
[rank0]:[titan] 2025-10-16 17:32:40,560 - root - INFO - step:  3  loss:  5.9824  grad_norm:  3.5676  memory:  5.82GiB(6.12%)  tps: 86,885  tflops: 35.33  mfu: 3.57%
[rank0]:[titan] 2025-10-16 17:32:40,746 - root - INFO - step:  4  loss:  5.1610  grad_norm:  2.7867  memory:  5.89GiB(6.21%)  tps: 88,525  tflops: 36.00  mfu: 3.64%
[rank0]:[titan] 2025-10-16 17:32:40,936 - root - INFO - step:  5  loss:  4.7838  grad_norm:  2.4660  memory:  6.23GiB(6.56%)  tps: 86,351  tflops: 35.11  mfu: 3.55%
[rank0]:[titan] 2025-10-16 17:32:41,127 - root - INFO - step:  6  loss:  4.5567  grad_norm:  2.4021  memory:  6.23GiB(6.56%)  tps: 86,018  tflops: 34.98  mfu: 3.54%
[rank0]:[titan] 2025-10-16 17:32:41,322 - root - INFO - step:  7  loss:  4.4087  grad_norm:  2.3600  memory:  6.23GiB(6.56%)  tps: 84,345  tflops: 34.30  mfu: 3.47%
[rank0]:[titan] 2025-10-16 17:32:41,520 - root - INFO - step:  8  loss:  4.3251  grad_norm:  2.2613  memory:  6.89GiB(7.26%)  tps: 82,943  tflops: 33.73  mfu: 3.41%
[rank0]:[titan] 2025-10-16 17:32:41,706 - root - INFO - step:  9  loss:  4.3709  grad_norm:  2.0616  memory:  6.89GiB(7.26%)  tps: 88,325  tflops: 35.92  mfu: 3.63%
[rank0]:[titan] 2025-10-16 17:32:41,896 - root - INFO - step: 10  loss:  4.2593  grad_norm:  2.0684  memory:  6.89GiB(7.26%)  tps: 86,348  tflops: 35.11  mfu: 3.55%
[rank0]:[titan] 2025-10-16 17:32:41,896 - root - INFO - Sleeping 2 seconds for other ranks to complete
[rank0]:[titan] 2025-10-16 17:32:43,896 - root - INFO - Training completed
[rank0]:[titan] 2025-10-16 17:32:47,371 - root - INFO - Process group destroyed
[rank0]:[rank0]:[W1016 17:32:47.493710282 ProcessGroup.hpp:940] Warning: No backend of type 0 found for Process Group with name undefined. Assuming no hooks are registered. (function hasHooks)

### CODE REVIEW

File: torchtitan/experiments/torchcomms/parallel_dims.py [hunk 1]
```diff
@@ -134,25 +178,121 @@ def _build_mesh_without_ep(self) -> DeviceMesh:
...
         self.comms = [*comm_per_dim.values(), comm]
+        return device_mesh
+
+    def _build_mesh_with_ep(self) -> DeviceMesh:
+        # With ep, dp_shard and ep are derived submeshes:
+        # dp_shard = dp_shard_mod_ep * dp_shard_in_ep
+        if self.etp == self.tp:
+            # ep = dp_shard_in_ep * cp
+            dp_shard_mod_ep = self.dp_shard * self.cp // self.ep
+            dp_shard_in_ep = self.ep // self.cp
```
Comment (tianyu-l) [inline torchtitan/experiments/torchcomms/parallel_dims.py:218]: This is workaround, assuming ep >= cp and ep < dp_shard * cp.
With new implementation https://github.com/pytorch/torchtitan/pull/1660, I expect these restrictions to be lifted. Do you think we could use new implementations?

Comment (mori360) [reply torchtitan/experiments/torchcomms/parallel_dims.py:218]: Thanks for the reference, it would be great to combine 2 build_mesh together.
The current design in this PR is to shape the mesh as detail as possibile in a single main communicator, and flatten the continuous dinmensions.
For https://github.com/pytorch/torchtitan/pull/1660, if we view communicators in 2 different shape, one for non-ep and one for ep, then could follow the similar path here.
@fduwjj what do you think of it?

Comment (mori360) [reply torchtitan/experiments/torchcomms/parallel_dims.py:218]: Discussed offline, will apply the new implementation after https://github.com/pytorch/torchtitan/pull/1660 landing.

File: torchtitan/experiments/torchcomms/parallel_dims.py [hunk 2]
```diff
@@ -49,81 +49,125 @@ def _calculate_ranks_per_dimension(
...
-            ["pp", "dp_replicate", "dp_shard", "cp", "tp"],
-        ):
-            if d > 1:
-                dims.append(d)
-                names.append(name)
+def _create_device_mesh(
+    world_size: int,
+    mesh_shape: tuple,
+    mesh_dim_names: List[str],
+) -> tuple:
```
Comment (fduwjj) [inline torchtitan/experiments/torchcomms/parallel_dims.py]: A better way to return here is:
Option1: return a dict so key will be str and value will be whatever you are returning here.
Option2: return an object

Directly returning tuple is hacky..

File: torchtitan/experiments/torchcomms/parallel_dims.py [hunk 3]
```diff
@@ -134,25 +178,121 @@ def _build_mesh_without_ep(self) -> DeviceMesh:
...
+
+        _flatten_comms(
+            flatten_ranks_per_dim,
+            comm,
+            flatten_mesh_dim_names,
+            device_mesh,
+            comm_per_dim,
+        )
+
+        # call .finalize() later to release the sub comm before the root comm
```
Comment (fduwjj) [inline torchtitan/experiments/torchcomms/parallel_dims.py]: please add comment on where this is called? for example just say inside trainer.py? (Of course, with details)

File: torchtitan/experiments/torchcomms/parallel_dims.py [hunk 4]
```diff
@@ -49,81 +49,138 @@ def _calculate_ranks_per_dimension(
...
+            _global_comm=comm,
+        )
+    except TypeError as e:
+        # TODO: remove this once PT 2.10 is released
+        if "_rank" in str(e):
+            for sub_comm in comm_per_dim.values():
+                sub_comm.finalize()
+            comm.finalize()
+            return {
+                "comm": None,
```
Comment (fduwjj) [inline torchtitan/experiments/torchcomms/parallel_dims.py]: shall we just directly return an empty dict?

File: torchtitan/experiments/torchcomms/parallel_dims.py [hunk 5]
```diff
@@ -49,81 +49,138 @@ def _calculate_ranks_per_dimension(
...
-            name="comms_test_n_d_parallel",
-        )
 
-        # Get current rank to determine which groups this rank belongs to
-        cur_rank = comm.get_rank()
+        result = _create_device_mesh(self.world_size, mesh_shape, mesh_dim_names)
+        comm = result["comm"]
+        device_mesh = result["device_mesh"]
+        mesh = result["mesh"]
+        comm_per_dim = result["comm_per_dim"]
```
Comment (fduwjj) [inline torchtitan/experiments/torchcomms/parallel_dims.py]: here you can use get("comm", None)

File: torchtitan/experiments/torchcomms/parallel_dims.py [hunk 6]
```diff
@@ -49,81 +49,138 @@ def _calculate_ranks_per_dimension(
...
-            if "_rank" in str(e):
-                for sub_comm in comm_per_dim.values():
-                    sub_comm.finalize()
-                comm.finalize()
-                return
-            raise
+        if device_mesh is None:
+            return None
+
+        cur_rank = comm.get_rank()
```
Comment (fduwjj) [inline torchtitan/experiments/torchcomms/parallel_dims.py:176]: Also add assert before L171 so that comm_per_dim, mesh and comm are not None.
